Add a "When to use Flow" section to the docs (#5793)
grafana/agent: docs/sources/flow/when-to-use-flow/opentelemetry-collector.md

Author: I'd like to add a Venn diagram such as this one in the beginning of this page:

![Agent and Collector Venn diagram](https://github.com/grafana/agent/assets/5834788/37dfed90-1cb8-4f5d-840d-05c607f9dff0)

---
description: Learn how to Flow compares to the OpenTelemetry Collector
menuTitle: OpenTelemetry Collector
title: OpenTelemetry Collector
weight: 100
---

# Comparing Flow with the OpenTelemetry Collector

## Which one of the two should you use? 

Flow may be a better choice if:
* You prefer River syntax over YAML.
* You would like to collect "profile" telemetry - Flow can do this via Pyroscope components.
* You would like to process Prometheus or Loki telemetry without converting it to OTLP.

The Collector may be a better choice if:
* You are already using the Collector, and do not need any of the features of Flow.
* You prefer YAML syntax over River.

> Below you can find more details on similarities and differences between the Collector and Flow.

Reviewer: I would remove this sentence.

## Similarities

### Many Collector components are available in Flow

Flow contains many of the components available inside the Collector:
* For example, the OTLP exporter is available as `otelcol.exporter.otlp`.
* The Grafana Agent development team is actively working on adding more of the Collector's components to Flow.

> * If a Flow user needs a Collector feature to be made available in the Agent, the Agent development team could work on implementing it.

Reviewer: These aren't components available inside the Collector.

I'd suggest we convert the bullets into sentences and just make a paragraph out of this information.

Author: I agree! I reworded it and it looks better to me now.

### Similar performance when processing OpenTelemetry signals natively

Most of Flow's `otelcol` components are just thin wrappers over a Collector component.

> Hence, the CPU and memory performance of Flow is similar.

Reviewer: I'd suggest we avoid using "hence" (not commonly used in American English).... Instead we can use something like "As a result".

## Differences

### Configuration language

> Collector is configured using yaml, whereas Flow is configured using River.

Reviewer: ```suggestion
Collector uses YAML for its configuration, and Flow uses River for its configuration.
```

#### Example - coalesce

> One of River's main advantages is its standard library. It contains handy functions such as coalesce, 
> which could be used to retrieve the first argument which is not null or empty:

Reviewer: ```suggestion
One of River's main advantages is its standard library. It contains handy functions such as coalesce.
You can use this function to retrieve the first argument which is not null or empty:
```

```river
```

### Modularity

> The Agent configuration is more flexible, modular, and allows for more opportunities to chain components together in a pipeline. 

Reviewer: ```suggestion
The Agent configuration is more flexible and modular, allowing you to chain more components together in a pipeline. 
```

I'm not sure if I'm changing the meaning of the sentence here.  Do you want  to focus on "more opportunities" to chain different sets of components into a single pipeline... or the ability to link together more components as a single set?

Reviewer: That adds a LOT more information. :-)

#### Example - retrieving data from a file

> Let's say you would like to use OAuth2 authentication in the Collector. If you need to retrieve `client_id` 

Reviewer: ```suggestion
Let's say you want to use OAuth2 authentication in the Collector. If you need to retrieve `client_id` 
```

> or `client_secret` from a file, then you would have to use the `client_id_file` or `client_secret_file` config parameters.

Reviewer: ```suggestion
or `client_secret` from a file, then you must use the `client_id_file` or `client_secret_file` config parameters.
```

```yaml
```

In the Agent, you'd use `otelcol.auth.oauth2` with the normal `client_id` and `client_secret` parameters, 

> and you would setup another component which retrieves those from a `local.file` component. 

Reviewer: ```suggestion
and you would set up another component that retrieves those from a `local.file` component. 
```

```river
```

> Moreover, the string could also come from a `remote.s3`, `remote.vault`, or `remote.http`. 

Reviewer: ```suggestion
The string could also come from a `remote.s3`, `remote.vault`, or `remote.http`. 
```

This gives Flow users lots of flexibility because strings coming out of those components 

> could be used for any parameter, in any component which requires a string - not just for 

Reviewer: ```suggestion
could be used for any parameter in any component which requires a string - not just for 
```

a `client_id` for OAuth2.

```river
```

### Flow can process Prometheus signals natively

> Collector needs to convert Prometheus signals to the OTLP format in order to process them.

Reviewer: ```suggestion
Collector needs to convert Prometheus signals to the OTLP format to process them.
```

Flow, on the other hand, can process those signals natively using components such as `prometheus.relabel`, `prometheus.relabel`, and `prometheus.remote_write`.
This could lead to better performance and ease of use.

### Flow documentation is consistent and structured

> Flow components tend to be documented in a more consistent way than Collector components. 

Reviewer: ```suggestion
Flow components are documented in a consistent ans structured way.
```

I'd suggest that we just stick with clear statements about Flow where possible.

### Some Collector features are not available in the Agent and vice-versa

> Agent doesn't have all the components which the Collector has. However, the Grafana  Agent development team is working on 
> adding new components all the time and we would be happy to add new components which Flow users need.

Reviewer: ```suggestion
Agent doesn't have all the components which the Collector has. However, the Grafana  Agent development team is constantly working on 
adding new components, and we would be happy to add new components that Flow users need.
```

### Flow supports "profile" telemetry signals

> OpenTelemetry currently does not support "profile" signals. Flow supports them through components such as `pyroscope.scrape` and `pyroscope.ebpf`.

Reviewer: ```suggestion
OpenTelemetry currently does not support "profile" signals. Flow supports "profile" signals through components such as `pyroscope.scrape` and `pyroscope.ebpf`.
```

---
description: Learn how to Flow compares to the OpenTelemetry Collector
menuTitle: OpenTelemetry Collector
title: OpenTelemetry Collector
weight: 100
---

# Comparing Flow with the OpenTelemetry Collector

## Which one of the two should you use? 

Flow may be a better choice if:
* You prefer River syntax over YAML.
* You would like to collect "profile" telemetry - Flow can do this via Pyroscope components.

> * You would like to process Prometheus or Loki telemetry without converting it to OTLP.

Reviewer 2: Is this the OSS documentation? One extra advantage of Flow is that Grafana Labs offers support for it.

The Collector may be a better choice if:
* You are already using the Collector, and do not need any of the features of Flow.

> * You prefer YAML syntax over River.

Reviewer 2: * You have a custom component, which you can incorporate in your own Collector distribution via the Collector Builder

Author: Is the main point the fact that you can add a custom component? Or is using the Collector Builder the advantage?

Either way, there is nothing stopping someone from forking the Agent and adding their own custom component. We just don't document that process very well, because in the long term we want to solve by developing a plugin-based system which doesn't require a new distribution/fork when a custom component is needed.

Reviewer 2: The use case is having a custom component, having a supported (and documented?) way to do that is what determines the tooling.

  a Git repository (`module.git`), or a HTTP endpoint (`module.http`).

#### Example - retrieving data from a file

Let's say you want to use OAuth2 authentication in the Collector. If you need to retrieve `client_id` 
or `client_secret` from a file, then you must use the `client_id_file` or `client_secret_file` config parameters.

```yaml
```

In the Agent, you'd use `otelcol.auth.oauth2` with the normal `client_id` and `client_secret` parameters, 
and you would set up another component that retrieves those from a `local.file` component. 

```river
```

The string could also come from a `remote.s3`, `remote.vault`, or `remote.http`. 
This gives Flow users lots of flexibility because strings coming out of those components 
could be used for any parameter in any component which requires a string - not just for 
a `client_id` for OAuth2.

```river
```

### Flow can process Prometheus signals natively

Collector needs to convert Prometheus signals to the OTLP format to process them.
Flow, on the other hand, can process those signals natively using components such as `prometheus.relabel`, `prometheus.relabel`, and `prometheus.remote_write`.

> This could lead to better performance and ease of use.

Reviewer 2: Our tests have shown that this leads to considerable performance gains.

@sh0rez , is there anything public we can link to?
